echemdb/data.py

```diff
@@ -1,31 +1,125 @@
 import os.path
 
-datadir = os.path.normpath(os.path.join(os.path.dirname(__file__), '..', 'data'))
+import hashlib
+
+datadir = os.path.normpath(os.path.join(os.path.dirname(__file__), '..', 'literature'))
 
 def collect_datapackages(data = datadir):
     # Collect all datapackage descriptors, see
     # https://specs.frictionlessdata.io/data-package/#metadata
     import os.path
     from glob import glob
-    descriptors = glob(os.path.join(data, '**', 'datapackage.json'), recursive=True)
+
+    ##NH  Here we want to build the datapackages from literature data on the fly
+    ## Note data could also be present already as datapackage.json or zip files
+    descriptors = glob(os.path.join(data, '**', '*.json'), recursive=True)
 
     # Read the package descriptors (does not read the actual data CSVs)
     from datapackage import Package
     packages = [Package(descriptor) for descriptor in descriptors]
 
     return packages
 
+
+def echemdb_id_from_descriptor(descriptor):
+    '''
+    We want to have a unique echemdb-id otherwise things look nasty
+    Let's use material-surface-uniqueid
+
+    e.g. namestr = descriptor['resources'][0]["name"] returns namestr = "briega-martos_2021_cation_XXX_p2_Fig1_Cs_2.3"
+    then we can do
+    '''
+    namestr = descriptor['resources'][0]["name"]
+    idend = hashlib.md5(namestr.encode()).hexdigest()[:7]
+    mat = property_reader(descriptor, "electrode material")
+    surf = property_reader(descriptor, "surface")
+
+    return "-".join([mat, surf, idend])
+
+
+
+
+
+
+
+
+def property_reader(descriptor, basic_property_name):
+    """
+    This function serves as a translation between basic property tags e.g.
+    ['electrode material', 'surface', 'electrolyte', 'reference', 'echemdb-id', 'path', 'relpath']
+    and how this is stored in datapackage descriptor
+
+    ['electrode material', 'surface', 'electrolyte', 'reference', 'echemdb-id', 'path', 'relpath']
+
+
+    """
+    def _finditem(obj, key):
+        if key in obj: return obj[key]
+        for k, v in obj.items():
+            if isinstance(v, dict):
+                item = _finditem(v, key)
+                if item is not None:
+                    return item
+
+
+
+    if basic_property_name == 'electrolyte':
+        els = []
+        for r in descriptor['electrochemical system']['electrolyte']['components']:
+            els.append(r["name"])
+        els.sort()
+        response = ", ".join(els)
```

Comment:
```suggestion
        components = [component["name"] for component in descriptor["electrochemical system"]["electrolyte"]["components"]]
        return ", ".join(sorted(components))
```
I believe that a list comprehension is much easier to read here than explicit looping.